python/amici/ode_export.py

```diff
@@ -1846,19 +1882,72 @@ def _compute_equation(self, name: str) -> None:
                 + self.eq('drootdt')
             )
 
-        elif name == 'stau':
+        elif name == 'deltax':
+            # fill boluses for Heaviside functions, as empty state updates
+            # would cause problems when writing the function file later
+            event_eqs = []
+            for event in self._events:
+                if event._state_update is None:
+                    event_eqs.append(sp.zeros(self.num_states_solver(), 1))
+                else:
+                    event_eqs.append(event._state_update)
+
+            self._eqs[name] = event_eqs
+
+        elif name == 'ddeltaxdx':
             self._eqs[name] = [
-                -self.eq('sroot')[ie, :] / self.eq('drootdt_total')[ie]
+                smart_jacobian(self.eq('deltax')[ie], self.sym('x'))
                 for ie in range(self.num_events())
             ]
 
-        elif name == 'deltasx':
+        elif name == 'ddeltaxdt':
+            self._eqs[name] = [
+                smart_jacobian(self.eq('deltax')[ie], time_symbol)
+                for ie in range(self.num_events())
+            ]
+
+        elif name == 'ddeltaxdp':
+            self._eqs[name] = [
+                smart_jacobian(self.eq('deltax')[ie], self.sym('p'))
+                for ie in range(self.num_events())
+            ]
+
+        elif name == 'stau':
             self._eqs[name] = [
-                smart_multiply((self.eq('xdot_old') - self.eq('xdot')),
-                               self.eq('stau')[ie])
+                -self.eq('sroot')[ie, :] / self.eq('drootdt_total')[ie]
                 for ie in range(self.num_events())
             ]
 
+        elif name == 'deltasx':
+            event_eqs = []
+            for ie, event in enumerate(self._events):
+                if event._state_update is not None:
+                    # ====== chain rule for the state variables ===============
+                    # get xdot with expressions back-substituted
+                    tmp_eq = smart_multiply(
+                        (self.sym('xdot_old') - self.sym('xdot')),
```

Comment:
sorry, I think this was an error on my end, this actually has to be `self.eq('xdot')`, not `self.sym('xdot')`, this may also solve the issues with nested events.

Reply:
I used to have "eq" here, and this was then wrong for nested events, also on the level of state variables, not only sensitivities... So no, I guess `sym` is definitely more correct than `eq`.

Reply:
I just see that you suggested multiple changes... Can try, whether all of them together will change things...
Let's give it a try and see...

Reply:
Arg, nope. Just implemented your suggestions and reran the tests. Both event test models fail now...
So: undo?

Reply:
Ah, yes. Looked at the matlab implementation and that also uses the strsyms. I think this has something to do at which point this difference is computed, which makes sense that this becomes relevant for nested events.